### PR TITLE
TUP-4352: new dynamic schema handling convention

### PR DESCRIPTION
**Pull request by stavytskyi**
I've created the pull request for storing our main discussion here. If we have discussion in Skype I will add summary here. It is easy to ask here question that related to code.

### CODE REVIEW

File: main/plugins/org.talend.designer.codegen/jet_stub/generic/component_begin.javajet [hunk 1]
```diff
@@ -137,22 +101,20 @@ org.talend.components.api.container.RuntimeContainer container_<%=cid%> = new or
...
-
-if (metadata != null) {
-	if (def instanceof InputComponentDefinition) {
+if(metadata != null){
 %>
-		org.talend.components.api.component.runtime.Source source_<%=cid%> = (org.talend.components.api.component.runtime.Source)sourceOrSink_<%=cid%>;
+	<%
+	if(def instanceof InputComponentDefinition){
+	%>
+		<%//take care SourceOrSink.validate will change the schema if it contains include-all-fields, so need to get design Avro schema before it%>
```
Comment (sgandon) [inline main/plugins/org.talend.designer.codegen/jet_stub/generic/component_begin.javajet]: here again, *validate* should never change the schema, this was a bug and was fixed.

**Comment by RyanSkraba** [comment]
Hello!  I created a new JIRA to deal with this problem... You are absolutely correct -- the schema passed to the output component *must* be the run-time schema (with all discovered columns and type information) and not the design time schema.  I'm not sure how this slipped by me earlier!  

(The original JIRA TCOMP-172 was focused on permitting components that don't contain a schema at all, but rely 100% on the runtime schema, which is related).

However, I'm not sure I understand all of the changes that you made in the javajet here... 

I'd like to propose an alternate solution at the branch:

https://github.com/Talend/tdi-studio-se/tree/feature/TCOMP-175-runtime-output-records
https://github.com/Talend/daikon/tree/feature/TCOMP-175-runtime-output-records

Can you take a look and see if it looks like the right direction?  Right now, it makes the assumption that all dynamic columns are String types, but I'll add the rest as soon as I validate this... There are a couple of NullPointerExceptions to iron out.

**Comment by sgandon** [comment]
This looks fine by me.
if @stavytskyi is ok with it we could close this PR and accept Ryan's job.

**Comment by RyanSkraba** [comment]
Hello!  I've got a working implementation at https://github.com/Talend/daikon/pull/29 (and it's related branch at the repo https://github.com/Talend/tdi-studio-se/pull/370).

I'll merge them today if there aren't any blocking comments.

I still need to make a pass through to fully support all dynamic column configurations, but I've tested String, numeric types and boolean.

**Comment by RyanSkraba** [comment]
Hello! Since the other implementation was implemented and merged, I'm closing this pull request. 

See https://github.com/Talend/daikon/pull/29

**Pull request closed without merging**